# T-DNAreader: Fast and precise identification of T-DNA insertion sites in plant genomes using RNA-sequencing data
  
![Image](https://github.com/user-attachments/assets/19bcfe30-5263-4a20-a16a-eff8704c646a)

For any question about TDNAreader, please contact miso5103@snu.ac.kr

## Dependencies
python3 >= 3.8

Bowtie2 >= 2.5.1 (https://github.com/BenLangmead/bowtie2)

Samtools >= 1.17 (http://www.htslib.org)

Trim_galore >= 0.6.10 (https://github.com/FelixKrueger/TrimGalore)

Deeptools >= 3.5.1 (https://deeptools.readthedocs.io/en/develop/content/installation.html)

GNU parallel (https://www.gnu.org/software/parallel/)

Required packages: re, io, pathlib, subprocess, argparse, time, multiprocessing, pandas, numpy

## Installation
Github install

```
git clone https://github.com/CDL-HongwooLee/TDNAreader
```

# Quick start

- TDNAreader are compatible with both FASTQ and BAM files.
- The BAM (SAM) files generated by STAR, Bowtie2 (with --local), BWA or other local sequencing aligners are acceptable.
- **Note: For paired-end sequencing, forward and reverse FASTQ files are separately aligned to the plant genome (treating them as single-end sequencing)
- Bowtie2 index for the provided TDNA.fa is required. (can be generated by ```bowtie2-build```)



### Quick run with the default parameters

For single-end sequencing, FASTQ input
```
python3 ../TDNAreader.py -fq sample_rep1.fastq.gz,sample_rep2.fastq.gz,sample_rep3.fastq.gz -ctl control_rep1.fastq.gz,control_rep2.fastq.gz,control_rep3.fastq.gz -o ./ -bed Annotation.bed -gtf Annotation.gtf -g reference_starPrefix -tdna TDNA.fa -p Prefix -bl blacklist.bed -t 12 --gzip
```

For single-end sqeuencing, BAM input
```
python3 TDNAreader.py -bam sample_rep1.bam,sample_rep2.bam,sample_rep3.bam -o ./ -bed Annotation.bed -gtf Annotation.gtf -tdna TDNA.fa -p Prefix -t 12 -bl blacklist.bed
```

For paired-end sequencing, FASTQ input
```
python3 ../TDNAreader.py -fq sample_rep1_1.fastq.gz:sample_rep1_2.fastq.gz,sample_rep2_1.fastq.gz:sample_rep2_2.fastq.gz,sample_rep3_1.fastq.gz:sample_rep3_2.fastq.gz -ctl control_rep1_1.fastq.gz:control_rep1_2.fastq.gz,control_rep2_1.fastq.gz:control_rep1_2.fastq.gz,control_rep3_1.fastq.gz:control_rep3_2.fastq.gz -o ./ -bed Annotation.bed -gtf Annotation.gtf -g reference_starPrefix -tdna TDNA.fa -p Prefix -bl blacklist.bed -t 12 --gzip --paired
```

For paired-end sequencing,
```
python3 TDNAreader.py -bam sample_rep1_1.bam:sample_rep1_2.bam,sample_rep2_1.bam:sample_rep2_2.bam,sample_rep3_1.bam:sample_rep3_2.bam -ctl control_rep1_1.bam:control_rep1_2.bam,control_rep2_1.bam:control_rep2_2.bam,control_rep3_1.bam:control_rep3_2.bam -o ./ -bed Annotation.bed -gtf Annotation.gtf -tdna TDNAfa -p Prefix -bl blacklist.bed -t 12 --paired
```

## TDNAreader.py

### Usage

```
usage: TDNAreader.py [-h] [-fq FASTQ] [-bam BAM] [-ctl CONTROL] [-o OUTDIR] -p PREFIX -tdna TDNA [-g GENOME] [-bed BED] [-gtf GTF] [-bl BLACKLIST] [-r RATIO] [-i INTRON] [-bL BOWTIEL] [-mapq MAPQ] [-d D]
                     [-l1 L1] [-l2 L2] [-m M] [-w WINDOW] [-x WEIGHT] [-u THRESHOLD] [-t THREAD] [--paired] [--gzip] [--wgs] [--discordant] [--tmp]

    T-DNAreader is a bioinformatics tool designed to identify T-DNA insertion sites (TISs) in plant genomes using NGS data.
    Contact address: miso5103@snu.ac.kr


optional arguments:
  -h, --help     show this help message and exit
  -fq FASTQ      Input FASTQ files. Gzipped files are allowed with --gzip.
                 For multiple replicates, separate files with commas (e.g. rep1.fastq,rep2.fastq).
                 For paired-end, separate pairs with colons (e.g. rep1_1.fastq:rep1_2.fastq,rep2_1.fastq:rep2_2.fastq).

  -bam BAM       Input BAM or SAM files.
                 Separate multiple files with commas (e.g. rep1.bam,rep2.bam).
                 For paired-end, separate pairs with colons (e.g. rep1_1.bam:rep1_2.bam).
                 Cannot be used with --discordant

  -ctl CONTROL   (Optional) Control (WT) FASTQ or BAM files (same format as -fq/-bam).
                 Use commas to separate replicates and colons for paired-end files.

  -o OUTDIR      Output directory where all output files will be saved.
                 Default = './'

  -p PREFIX      Prefix for naming all output files.

  -tdna TDNA     Path to T-DNA FASTA (LB to RB).
                 Requires a pre-built Bowtie2 index.
                 Exclude vector backbone sequences outside LB-RB.

  -g GENOME      Prefix for your genome index:
                 - STAR: path to genomeDir
                 - Bowtie2: path prefix to .bt2 files
                 (e.g., /path/to/genomeDir or /path/to/bowtie2/index_prefix).

  -bed BED       (Optional) BED6 file with genomic features.
                 Column 4 should contain gene IDs.
                 Overlapping TISs will be annotated.

  -gtf GTF       (Optional) GTF file for STAR alignment and gene annotation.

  -bl BLACKLIST  (Optional) BED file of regions to exclude.
                 If T-DNA contain endogenous genomic sequences,
                 it is strongly recommended to specify the those positions to avoid misinterpretation.

  -r RATIO       Minimum fraction of each read that must align to the genome (STAR: --outFilter*OverLread).
                 Default = 0.33 (recommended for 125-150 bp).
                 Use ~0.5 for 80-125 bp, or
                 0.6 - 0.7 is recommend for 50-80 bp.
                 Only work for RNA-seq.

  -i INTRON      (Optional) Maximum intron size for STAR (--alignIntronMax).
                 Default = 30000

  -bL BOWTIEL    (Optional) Seed length for Bowtie2 local alignment (bp).
                 Must be >3 and <= '-l1'.
                 Default = 10

  -mapq MAPQ     (Optional) Mapping quality threshold for WGS analysis.
                 Default = 30

  -d D           (Optional) Distance (bp) from T-DNA borders within which to apply the â€œ-l1â€ matchâ€length threshold.
                 Reads mapped within this distance from T-DNA borders used the threshold specified by -l1.
                 Reads mapped beyond this distance use the threshold specified by l2.
                 Default = 200 (100<=d<=300 is recommended)

  -l1 L1         (Optional) Min matched sequence length (reads mapped close to T-DNA borders).
                 Higher increases stringency.
                 Default = 18 (15<=l1<=30 is recommended)

  -l2 L2         (Optional) Min matched sequence length (reads mapped in internal T-DNA regions).
                 Higher increases stringency.
                 Default = 30 (30<=l2<=50 is recommended)

  -m M           (Optional) Max mismatches allowed in T-DNA alignment.
                 Default = 1 (bp) (0<=m<=3 is recommended)

  -w WINDOW      (Optional) Merge distance for adjacent T-DNA insertion sites.
                 Default = 500 (bp)

  -x WEIGHT      (Optional) Comma-separated integers defining the scoring weights.
                 Weights for group1-5 reads.
                 Default = 700,300,300,1,1

  -u THRESHOLD   (Optional) Score threshold  to filter false positives.
                 Default = 1000

  -t THREAD      Number of threads to use.
                 Default = 1.

  --paired       Inputs are from paired-end data.
                 Separate pairs with ':'.

  --gzip         Specify if FASTQ inputs are gzipped.

  --wgs          (Optional) The input is whole-genome sequencing data.

  --discordant   (Optional) Include discordant pairs in scoring.
                 If specified, this increases runtime.

  --tmp          (Optional) Keep intermediate temporary files

```



## Input and output data

### Input

#### FASTQ files
FASTQ files are acceptable, and will be aligned using STAR or Bowtie2('--wgs').


#### BAM files
- The BAM (SAM) files generated by STAR, Bowtie2 (with --local), BWA or other local sequencing aligners are acceptable.
  - For STAR, use {name}.sortedByCoord.bam files.
- **Note: For paired-end sequencing, forward and reverse FASTQ files are separately aligned to the plant genome (treating them as single-end sequencing).
- (Optional) The modification of STAR alignment options may improve the sensitivity of TDNAreader. You can specify ```--outFilterMatchNminOverLread```, ```--outFilterScoreMinOverLread```, and ```--outFilterMismatchNoverLmax``` to an lower value (e.g., 0.33, 0.5) to identify more reads. The range from 0.33 to 0.5 is recommended depending on your read length.

#### tdna.fa

- FASTA file containing T-DNA sequences inserted into the plant genome: from left-border (LB) to right-border (RB)
- Sequences outside LB-RB regions should be excluded
- Bowtie2 index for this file is required prior to running TDNAreader (use ```bowtie2-build```)

### Output

Output file is generated as ```{outdir}/{prefix}.TDNA.txt```, ```{outdir}/{prefix}.read_info.txt```.

Example 
{prefix}.TDNA.txt
```
chrom	start	end	TISs	gene	T-DNA	junction	orientation	count	score	key	sample
Chr4	14016003	14016004	Chr4::14016003::R::+	AT4G28300	pROK2(SALK)::29	REF:T-DNA	+	47,1,1,0,0	33500	Chr4::14016003::R::+	1_floe1-1
```
- chrom: chromosome
- start: start position
- end: end position
- TISs: name of TISs including adjacent TIS. {chrom}::{position}::{L/R}::{orientation}
- gene: Overlapping genes
- T-DNA: name of TDNA (derived from tdna.fa)
- junction: Relative position of T-DNA. T-DNA:REF (L) or REF:T-DNA (R)
- orientation: the orientation of inserted T-DNA (+ or -)
- count: Read count (groups1-5)
- score: Weighted sum of read count.
- key: name of TIS
- sample: {prefix}_{replicate_n}-{F/R}


{prefix}.read_info.txt
```
chrom	start	end	readid	flag	cigar	junction	seq	tdna	tdna_flag	tdna_pos	tdna_cigar	matchedL	tdna_seq	prefix	TIS
Chr4	14015873	14016003	SRR13765633.1338395	0	131M19S	R	TCGTCGTCTCACTCTCAACATGGTGAGGACCGTGTCGCCACTCCTGTTCCAGAGCCTAAGAAGAGCGAGAACACCTCTGATGCACACAACCAGCAGCTTGCACTTGCTTTGCCTCACCAAATAGCCCCACATTGACGCTTAGACAACTTA	pROK2(SALK)	+	29	19M	19	TTGACGCTTAGACAACTTA	1_floe1-1_2-R	Chr4::14016003::R::+
Chr4	14015873	14016003	SRR13765634.8319149	16	131M19S	R	TCGTCGTCTCACTCTCAACATGGTGAGGACCGTGTCGCCACTCCTGTTCCAGAGCCTAAGAAGAGCGAGAACACCTCTGATGCACACAACCAGCAGCTTGCACTTGCTTTGCCTCACCAAATAGCCCCACATTGACGCTTAGACAACTTA	pROK2(SALK)	+	29	19M	19	TTGACGCTTAGACAACTTA	1_floe1-1_3-R	Chr4::14016003::R::+
Chr4	14015874	14016003	SRR13765633.11143970	0	130M20S	R	CGTCGTCTCACTCTCAACATGGTGAGGACCGTGTCGCCACTCCTGTTCCAGAGCCTAAGAAGAGCGAGAACACCTCTGATGCACACAACCAGCAGCTTGCACTTGCTTTGCCTCACCAAATAGCCCCACATTGACGCTTAGACAACTTAA	pROK2(SALK)	+	29	20M	20	TTGACGCTTAGACAACTTAA	1_floe1-1_2-R	Chr4::14016003::R::+
Chr4	14015875	14016003	SRR13765633.16534429	16	129M21S	R	GTCGTCTCACTCTCAACATGGTGAGGACCGTGTCGCCACTCCTGTTCCAGAGCCTAAGAAGAGCGAGAACACCTCTGATGCACACAACCAGCAGCTTGCACTTGCTTTGCCTCACCAAATAGCCCCACATTGACGCTTAGACAACTTAAT	pROK2(SALK)	+	29	21M	21	TTGACGCTTAGACAACTTAAT	1_floe1-1_2-F	Chr4::14016003::R::+
Chr4	14015875	14016003	SRR13765634.13245767	16	129M21S	R	GTCGTCTCACTCTCAACATGGTGAGGACCGTGTCGCCACTCCTGTTCCAGAGCCTAAGAAGAGCGAGAACACCTCTGATGCACACAACCAGCAGCTTGCACTTGCTTTGCCTCACCAAATAGCCCCACATTGACGCTTAGACAACTTAAT	pROK2(SALK)	+	29	21M	21	TTGACGCTTAGACAACTTAAT	1_floe1-1_3-R	Chr4::14016003::R::+
```
